File: app/src/main/java/com/example/buildmypc/ui/parts/parts/Part.java
package com.example.buildmypc.ui.parts.parts;

// the start of the part hierarchy
public class Part {

	private String model;
	private String manufacturer;

	public Part(String model, String manufacturer) {
		this.model = model;
		this.manufacturer = manufacturer;
	}

	public String getName() {
		return manufacturer + " " + model;
	}

	public String getModel() {
		return model;
	}

	public void setModel(String model) {
		this.model = model;
	}

	public String getManufacturer() {
		return manufacturer;
	}

	public void setManufacturer(String manufacturer) {
		this.manufacturer = manufacturer;
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (o == null || getClass() != o.getClass()) return false;

		Part part = (Part) o;

		if (getModel() != null ? !getModel().equals(part.getModel()) : part.getModel() != null)
			return false;
		return getManufacturer() != null ? getManufacturer().equals(part.getManufacturer()) : part.getManufacturer() == null;
	}

	@Override
	public int hashCode() {
		int result = getModel() != null ? getModel().hashCode() : 0;
		result = 31 * result + (getManufacturer() != null ? getManufacturer().hashCode() : 0);
		return result;
	}
}

File: app/src/main/java/com/example/buildmypc/ui/parts/parts/Part.kt
package com.example.buildmypc.ui.parts.parts

import android.os.Parcel
import android.os.Parcelable
import android.os.Parcelable.Creator

// the start of the part hierarchy
open class Part : Parcelable {
	var model: String?
	var manufacturer: String?

	constructor(model: String?, manufacturer: String?) {
		this.model = model
		this.manufacturer = manufacturer
	}

	constructor(`in`: Parcel) {
		model = `in`.readString()
		manufacturer = `in`.readString()
	}

	val name
		get() = "$manufacturer $model"

	// returns the base number of params -- will be overridden for every subclass of Part
	open val paramCount = 2

	override fun equals(other: Any?): Boolean {
		if (this === other) return true
		if (other == null || javaClass != other.javaClass) return false
		val part = other as Part
		if (if (model != null) model != part.model else part.model != null) return false
		return if (manufacturer != null) manufacturer == part.manufacturer else part.manufacturer == null
	}

	override fun hashCode() =
		31 * (if (model != null) model.hashCode() else 0) + if (manufacturer != null) manufacturer.hashCode() else 0

	override fun describeContents() = 0

	override fun writeToParcel(dest: Parcel, flags: Int) {
		dest.writeString(model)
		dest.writeString(manufacturer)
	}

	companion object {
		@JvmField
		val CREATOR = object : Creator<Part> {
			override fun createFromParcel(`in`: Parcel) = Part(`in`)

			override fun newArray(size: Int): Array<Part?> = arrayOfNulls(size)
		}
	}

	override fun toString() = "$manufacturer $model"
}